pubsub/google/cloud/pubsub_v1/subscriber/_protocol/leaser.py

```diff
@@ -0,0 +1,186 @@
+        repeats.
+        """
+        while self._subscriber.is_active and not self._stop_event.is_set():
+            # Determine the appropriate duration for the lease. This is
+            # based off of how long previous messages have taken to ack, with
+            # a sensible default and within the ranges allowed by Pub/Sub.
+            p99 = self._subscriber.ack_histogram.percentile(99)
+            _LOGGER.debug('The current p99 value is %d seconds.', p99)
+
+            # Make a copy of the leased messages. This is needed because it's
+            # possible for another thread to modify the dictionary while
+            # we're iterating over it.
+            leased_messages = copy.copy(self._leased_messages)
+
+            # Drop any leases that are well beyond max lease time. This
+            # ensures that in the event of a badly behaving actor, we can
+            # drop messages and allow Pub/Sub to resend them.
+            cutoff = (
+                time.time() -
+                self._subscriber.flow_control.max_lease_duration)
+            to_drop = [
+                requests.DropRequest(ack_id, item.size)
+                for ack_id, item
+                in six.iteritems(leased_messages)
+                if item.added_time < cutoff]
+
+            if to_drop:
+                _LOGGER.warning(
+                    'Dropping %s items because they were leased too long.',
+                    len(to_drop))
+                self._subscriber.drop(to_drop)
+
+            # Remove dropped items from our copy of the leased messages (they
+            # have already been removed from the real one by self.drop).
+            for item in to_drop:
+                leased_messages.pop(item.ack_id)
+
+            # Create a streaming pull request.
+            # We do not actually call `modify_ack_deadline` over and over
+            # because it is more efficient to make a single request.
+            ack_ids = list(leased_messages.keys())
+            if ack_ids:
+                _LOGGER.debug('Renewing lease for %d ack IDs.', len(ack_ids))
+
+                # NOTE: This may not work as expected if ``consumer.active``
+                #       has changed since we checked it. An implementation
+                #       without any sort of race condition would require a
+                #       way for ``send_request`` to fail when the consumer
+                #       is inactive.
+                self._subscriber.modify_ack_deadline([
+                    requests.ModAckRequest(ack_id, p99) for ack_id in ack_ids])
+
+            # Now wait an appropriate period of time and do this again.
+            #
+            # We determine the appropriate period of time based on a random
+            # period between 0 seconds and 90% of the lease. This use of
+            # jitter (http://bit.ly/2s2ekL7) helps decrease contention in cases
+            # where there are many clients.
+            snooze = random.uniform(0.0, p99 * 0.9)
```

Comment:
Randomizing the exact delay makes sense, but going to 0.0 is interesting instead of some small, but non-zero, number.